### PR TITLE
[READY] Added change_cloud nanite program

### PR DESCRIPTION
**Pull request by Blevruz**
## About The Pull Request

Adds a "change cloud" program, researched in Smart Nanite Programming. 
This program takes a cloud ID, then, when triggered, changes the nanite's cloud to the specified ID. 
This does not result in an immediate change in the nanites' programs, but the programs will change at the next backup.

## Why It's Good For The Game

Allows for more possibilities in nanite programming.

## Changelog
:cl:
add: Added a "change cloud" nanite program
/:cl:

### CODE REVIEW

File: code/modules/research/nanites/nanite_programs/utility.dm [hunk 1]
```diff
@@ -356,3 +356,19 @@
...
+	trigger_cost = 0
+	trigger_cooldown = 10
+	rogue_types = list(/datum/nanite_program/glitch)
+	
+/datum/nanite_program/change_cloud/register_extra_settings()
+	extra_settings[NES_CLOUD_OVERWRITE] = new /datum/nanite_extra_setting/number(0, 0, 100)
+
+/datum/nanite_program/change_cloud/on_trigger(comm_message)
+	var/datum/nanite_extra_setting/cloud = extra_settings[NES_CLOUD_OVERWRITE]
+	SEND_SIGNAL(host_mob, COMSIG_NANITE_SET_CLOUD, cloud.get_value() )
```
Comment (Rohesie) [inline code/modules/research/nanites/nanite_programs/utility.dm:374]: Signals need to happen at the base of the proc.

**Comment by XDTM** [comment]
Changing the cloud to 0 or something non-existent will make everyone's nanites decay and unlike other sabotages it's not reverted by fixing the cloud, which is pretty powerful.
You can achieve a similar effect by deleting the cloud or destroying the console, though, so it might be fine. 

It's also worth noting that this will allow powergamers to disable and enable their cloud sync at will, as long as the cloud switch program itself does not get corrupted. If that tactic becomes too overused it might require a faster nanite decay to balance it out.

File: code/modules/research/nanites/nanite_programs/utility.dm (hunk 1 above)
Comment (Blevruz) [reply code/modules/research/nanites/nanite_programs/utility.dm:374]: I see, replacing SEND_SIGNAL it with the set_cloud proc.

**Comment by moocowswag** [comment]
Sounds like a neat way to cycle through different modes when out of combat and breach the hard limit on the amount of programs you have.

Say you die, a death sensor triggers and switches your cloud id to 5, cloud id 5 is riddled with tons of health and damage sensors along with a brain damage program that turns itself on and off so its on just enough to keep your brain from decaying + minimum nanite rules to conserve the bare minimum nanites you need to finally defib when you go over 1% hp (just above soft crit)



The only problem is if this program doesnt have a corrupt variant, you could just have it on a 1 second restart timer so youll always be able to trigger it in the event an emp sat your cloud id to 0.

But then again that seems fair considering people have nerfed nanites saying that its weaknesses are obsolete, so they might aswell be.

**Comment by Blevruz** [comment]
I think the fact you can't force the nanites to synch is enough to keep it reasonable, but nanite decay is pretty slow in general. Could stand to be sped up.

I'm all for encouraging specialised clouds, as that is a more interesting way to avoid nanites being able to do everything all the time, but something would need to be done to ensure you can't just do every program in a single cloud. Maybe lowering the maximum size of a cloud?

It currently can get corrupted, so it is vulnerable to EMP. Perhaps the cooldown would need adjusting to avoid precisely this scenario?

**Comment by moocowswag** [comment]
You already cant do everything in a single cloud thanks to the changes to blood cleansing making it now require atleast 4 sensors to make them work properly without killing slime people.

But then again this program would make it possible to just have the public chamber connect people to a cloud that sorts them to other clouds based on their species and if they are fully augmented or not (with specie sensors and a button)

If that becomes meta you have effectively atleast doubled the amount of time someone would spend coding and interacting with nanites which is a huge leap towards making this feel more like an actual job that someone in the research department could dedicate their time to.

**Comment by Rohesie** [comment]
This has changes pending and went stale twice. Let us know if you are still interested in working on this, make the requested changes and, ideally, get the interest of a maintainer to re-open this.

**Comment by Blevruz** [comment]
> 
> 
> This has changes pending and went stale twice. Let us know if you are still interested in working on this, make the requested changes and, ideally, get the interest of a maintainer to re-open this.

Forgot to hit "resolved" on the requested changes. It is done now.

File: code/modules/research/nanites/nanite_programs/utility.dm (hunk 1 above)
Comment (Rohesie) [reply code/modules/research/nanites/nanite_programs/utility.dm:374]: This in still unaddressed.

**Review by Rohesie — CHANGES_REQUESTED** [review]

**Comment by Blevruz** [comment]
Does this change address it?

File: code/modules/research/nanites/nanite_programs/utility.dm (hunk 1 above)
Comment (Blevruz) [reply code/modules/research/nanites/nanite_programs/utility.dm]: This should be addressed now.

**Comment by Rohesie** [comment]
Yep.
Pending balance review.
See if there's a maintainer interested in this feature.

**Comment by Blevruz** [comment]
This feature seems to be too poorly balanced to implement, so I'll close this PR.

**Pull request closed without merging**